test/shared/src/main/scala/zio/test/Gen.scala

```diff
@@ -18,47 +18,99 @@ package zio.test
 case class Gen[-R, +A](sample: ZStream[R, Nothing, Sample[R, A]]) { self =>
-  final def <*>[R1 <: R, B](that: Gen[R1, B]): Gen[R1, (A, B)] = self zip that
 
-  final def flatMap[R1 <: R, B](f: A => Gen[R1, B]): Gen[R1, B] =
-    Gen(sample.flatMap(sample => f(sample.value).sample))
+  final def <*>[R1 <: R, B](that: Gen[R1, B]): Gen[R1, (A, B)] =
+    self zip that
 
-  final def map[B](f: A => B): Gen[R, B] = Gen(sample.map(_.map(f)))
+  final def filter(f: A => Boolean): Gen[R, A] =
+    flatMap(a => if (f(a)) Gen.const(a) else Gen.empty)
 
-  final def zip[R1 <: R, B](that: Gen[R1, B]): Gen[R1, (A, B)] = self.flatMap(a => that.map(b => (a, b)))
+  final def flatMap[R1 <: R, B](f: A => Gen[R1, B]): Gen[R1, B] = Gen {
+    sample.flatMap { a =>
+      f(a.value).sample.map { b =>
+        val rest = a.shrink.flatMap(a => f(a).sample).flatMap(sample => ZStream(sample.value) ++ sample.shrink)
+        Sample(b.value, b.shrink ++ rest)
+      }
+    }
+  }
+
+  final def map[B](f: A => B): Gen[R, B] =
+    Gen(sample.map(_.map(f)))
+
+  final def noShrink: Gen[R, A] =
+    Gen(sample.map(sample => Sample.noShrink(sample.value)))
+
+  final def withShrink[R1 <: R, A1 >: A](f: A => ZStream[R1, Nothing, A1]): Gen[R1, A1] =
+    Gen(sample.map(sample => Sample(sample.value, f(sample.value) ++ sample.shrink.flatMap(f))))
+
+  final def zip[R1 <: R, B](that: Gen[R1, B]): Gen[R1, (A, B)] =
+    self.flatMap(a => that.map(b => (a, b)))
+
+  final def zipAll[R1 <: R, B](that: Gen[R1, B]): Gen[R1, (Option[A], Option[B])] =
+    Gen {
+      (self.sample.map(Some(_)) ++ Stream(None).forever)
+        .zip(that.sample.map(Some(_)) ++ Stream(None).forever)
+        .collectWhile {
+          case (Some(a), Some(b)) => a.zipWith(b) { case (a, b) => (Some(a), Some(b)) }
+          case (Some(a), None)    => a.map(a => (Some(a), None))
+          case (None, Some(b))    => b.map(b => (None, Some(b)))
+        }
+    }
 
   final def zipWith[R1 <: R, B, C](that: Gen[R1, B])(f: (A, B) => C): Gen[R1, C] =
     (self zip that) map f.tupled
 }
+
 object Gen {
 
+  /**
+   * A generator of booleans.
+   */
+  final val boolean: Gen[Random, Boolean] =
+    choose(false, true)
+
+  final def char(range: Range): Gen[Random, Char] =
+    integral[Char](range)
+
+  final def choose[A](a: A, as: A*): Gen[Random, A] =
+    oneOf(const(a), as.map(a => const(a)): _*)
+
   /**
    * A constant generator of the specified value.
    */
-  final def const[A](a: => A): Gen[Any, A] = Gen(ZStream.succeedLazy(Sample.noShrink(a)))
+  final def const[A](a: => A): Gen[Any, A] =
+    Gen(ZStream.succeedLazy(Sample.noShrink(a)))
 
   /**
    * A constant generator of the specified sample.
    */
-  final def constSample[R, A](sample: => Sample[R, A]): Gen[R, A] = fromEffectSample(ZIO.succeedLazy(sample))
+  final def constSample[R, A](sample: => Sample[R, A]): Gen[R, A] =
+    fromEffectSample(ZIO.succeedLazy(sample))
+
+  final def double(min: Double, max: Double): Gen[Random, Double] =
```

Comment:
Useful to add `float`, `short`, `long`, etc. Maybe the bounded variants, but also: `anyDouble`, `anyChar`, `anyShort`, `anyLong`, `anyFloat`, etc.

Reply:
Yes. One potential issue with the bounded variants for types like `Long` and `Double` is that they are "bigger" than the `Int` that underlies range. I wonder if we should roll our own simple range for this so we can be consistent across numeric types or just use something like `min`, `max`, `step`.